### DISCUSSION TITLE
Deno vs Node for writing tests

### DISCUSSION THREAD

**Discussion by lgalabru** [stx-labs/clarinet#1022] [Polls]
We want to re-evaluate some design choices that we made when we started Clarinet.
Our initial rationale was to create a tool that would assist you writing clarity code, and some unit tests without having to deal with the usual dependency hell coming with node.
Embedding Deno in Clarinet is coming with a cost:
- Maintaining and updating this integration is time consuming
- A LOT of dependencies are being pulled in clarinet (final binary 90Mb, instead of 15Mb)
- Deno is a tedious project to compile, and it's incompatible with certain OS config (https://github.com/hirosystems/clarinet/issues/518)  
- We can't run clarinet tests in the browser - which would be great for developer tooling.

Switching to Node would let us address all the points above, but would come with a cost: developers would have to install node on their machine, if they want to run clarinet tests.

**Comment by scottsolowild** [comment]
To understand all options here: Is there a way to keep deno out of the final binary, say by compiling a separate version that's only used for development installs & testing (`clarinet-devel` for example)?

**Reply by hugoclrd** [reply]
I don't think so. Deno is fully part of the final product, it's used by Clarity developers to execute the Clarity tests

**Comment by diwakergupta** [comment]
> Switching to Node would let us address all the points above, but would come with a cost: developers would have to install node on their machine, if they want to run clarinet tests.

I think it's a safe assumption (and we can test it if needed) that the vast majority of developers using `clarinet` already have node installed. And even if they don't, it's not a crazy ask. So if that's the only downside, it's a fairly minor one IMO.

**Comment by sabbyanandan** [comment]
@lgalabru: What would it mean for existing Clarinet-generated projects if we migrated to Node? Assuming you'd gracefully handle migrating them over in a backward-compatible manner, I wanted to check how/if this would disrupt any existing setup.

**Reply by lgalabru** [reply]
It would be pretty disruptive, we could keep the same testing API, but a lot of things like imports etc would change.
I think we're talking about Clarinet 2.0 materials.

**Reply by hugoclrd** [reply]
It does look like a major upgrade. But we could include a Codemod to make things easier

**Comment by moodmosaic** [comment]
This comes with a cost (or rather *costs*) as you'll now have to educate the users of the software on how they can migrate their tests.

And you can't really touch existing tests; you'll make them less trustworthy.

So Deno tests MUST still be able to run.

Deno has a built-in test discovery mechanism. Node.js doesn't. A problem to be solved.

On the other hand, if you switch to Node you'll make the integration with fuzzing and property tests a whole lot easier.

Happy to discuss this in very detail.

**Reply by hugoclrd** [reply]
Interesting insights, thanks for sharing.

> you'll now have to educate the users of the software on how they can migrate their tests

The way I see it is that developers will be able to tests smart contracts with existing technologies and the test frameworks they already know. (You could test smart contract with the same framework as a frontend application like mocha or jest). Instead of learning clarinet test

> And you can't really touch existing tests; you'll make them less trustworthy.

I don't understand what you mean by that, could you briefly elaborate?

**Comment by moodmosaic** [comment]
> The way I see it is that developers will be able to tests smart contracts with existing technologies and the test frameworks they already know. (You could test smart contract with the same framework as a frontend application like mocha or jest). Instead of learning clarinet test

The question is, should Clarity developers think [in or *into*](https://codeblog.jonskeet.uk/2008/04/23/programming-quot-in-quot-a-language-vs-programming-quot-into-quot-a-language/) the Clarity language?

That was one of the reasons behind the success of [foundry](https://github.com/foundry-rs/foundry) on Ethereum; it minimizes the context switching between Solidity and JS/TS and focus on what matters most<sup>1</sup> (the smart contract language).

IMHO, the direction that @igorsyl, @jo-tm, @sergey-shandar, are taking in sBTC *is* the way to go;  tests are written in Clarity, and discovered and run via Clarinet using Deno.

That way, in sBTC, they get almost for free the ability to do [Invariant Testing](https://github.com/stacks-network/stacks-blockchain/discussions/3732#discussioncomment-6091365), and adding support for property tests (and fuzz tests afterwards) won't require any change inside Clarinet, Deno, (or Node.js in this case).

<sup>1</sup>Because (non-upgradable) smart contracts are like hardware devices; once they're shipped they can't really be changed.

**Reply by hugoclrd** [reply]
These are great insights!
@igorsyl, I'd like to set up a call with some people from the sBTC team to learn more about their approach regarding testing. Who should I include in the loop? It would be awesome if it can happen this week.
Let's see how we can support tests in Clarity and improve it 🙌 

cc @lgalabru

**Reply by igorsyl** [reply]
@hugocaillard : Sweet! We need to write the grant for the sBTC Testing Team (4 to 6 devs for 1 Quarter) which is due this Friday and we need all the help we can get! Book time here and I'll add the relevant folks!
https://savvycal.com/trust-machines-igor/chat

**Reply by igorsyl** [reply]
@hugocaillard I've had more thoughts about this, and I think we should have clarity files become LISP files with the following structure:

[Smart contract code written in Clarity, a decidable LISP DSL]
[Testing code written in turning complete LISP]

Then, when deploying, clarinet could strip out all non-Clarity code from the source code.
Now the problem is to find a good LISP runtime in Rust, or in the extreme case, use SBCL. We would not need to use Deno or Node.
Let me know if this sounds interesting, and I can do a proof of concept.

cc @kantai @lgalabru

**Reply by moodmosaic** [reply]
If the data for the arrange phase of a test is generated externally via TypeScript/fast-check you shouldn't need all this overhead. Unless I am missing something.

**Reply by igorsyl** [reply]
Agreed, but OP's initial concern included binary size. Arguably, adding TS is the overhead compared to a small LISP runtime.
Clarinet could be rebuilt using Rust and LISP only -- no TS. The arrange phase could be coded in either of these languages.

**Reply by moodmosaic** [reply]
Yes, since rust has proptest.

But then I wonder how the size/complexity/maintenance looks like wrt the backwards compatibility layer. To support all the existing test suites out there.

I'm not sure I understand the concern of the size of the deno, by the way.

At this point I think I'm missing important context. If someone can kindly point me to that context, I might be able to help.

**Reply by hugoclrd** [reply]
Hey @igorsyl @moodmosaic,

I'll try to clarify the context:

### What is the issue and why addressing it now
- Clarinet currently embeds deno_core@1.146.0 (and old version) to provide a JS/TS runtime to execute tests. It doesn't simply has Deno as a cargo dependency, but some of [Deno's source code is copy/pasted](https://github.com/hirosystems/clarinet/blob/edb77595c0f5e79ddacd0e9b5f4a0d6ec3cac776/components/clarinet-cli/build.rs#L1) into Clarinet (because of the way Deno is packaged)
- This Deno integration is hard to maintain (because of frequent changes in Deno's architecture), and even though today (as of deno_core@0.194.0) it seems a bit more easy to embed/maintain, it's still not ideal.
- Why now? deno_core@1.146.0 fails to build on Homebrew (for recent macos version) and breaks the release, so we have to find a solution if we want to keep releasing clarinet upgrades on Homebrew in the future.

### Solution(s)
- Ultimately, I don't think that it's the role of Clarinet to embed a JS runtime. It takes a lot of time to maintain and the Clarinet team could focus on other priorities.
- There exists better alternatives, such as exporting a testing lib that allow to interact with a clarity session, call contracts, makes assertions.
- This testing lib _could_ be used in Rust, but we think it's better/easier for developers to use JS (_instead of Rust, I'm not talking about other languages here_) to write tests. So this testing lib can be exposed through WASM.
- Plus, developers can take advantage of Stacks.js to interact with clarity values (instead of the current custom way of doing it with Clarinet).
- Clarinet won't have to embed any other language runtime. Instead, the testing lib will be available in any environment  supporting WASM
- We'll (most likely) recommend Node.js to run tests, as a bonus, @moodmosaic I'll connect with you to see how we can enable fuzz testing. I think this new approach will give a lot more flexibility and possibilities to improve the testing experience.
- 👉  **More context about this current approach in this issue #1065** 
- Let me know what you think about it, we need as many feedback as possible on this. I'll also open a thread on Discord about that.

--- 

- As a side effect, we can remove Deno and all of its dependencies from Clarinet. It means faster build time, smaller binary output (although it's not a concern, just a bonus side effect), and less maintenance. And will unlock the releases on Homebrew.


@igorsyl 
I feel like the Clarity/Lisp discussion is bigger than this topic, and maybe should happen somewhere else (with the blockchain team), what do you think?

**Reply by igorsyl** [reply]
> I feel like the Clarity/Lisp discussion is bigger than this topic, and maybe should happen somewhere else (with the blockchain team), what do you think?

@hugocaillard : Agreed, I think a discussion in the stacks-blockchain repo is needed.

**Reply by moodmosaic** [reply]
@hugocaillard, thank you for the above detailed points 👍 

* Deno has been designed with security as one of its primary focuses. Some of the security features that make Deno stand out compared to Node.js are the Sandboxed environment and Fine-grained Permissions.
* Node.js was not designed with the same emphasis on security. Its ecosystem and package manager, npm, are historically known for issues related to dependencies, where malicious packages or dependency chains could sometimes lead to security vulnerabilities.

That's a rather expensive tradeoff 🤔

(OTOH, security also relies on best practices, code quality, and how the software is maintained.)

---

Perhaps, if you decide to use Node.js you could maintain the original Clarinet/Deno syntax `Clarinet.test(...` to keep all existing test-suites running...

**Reply by hugoclrd** [reply]
@moodmosaic 
Since the core of the lib will be WASM, it will be usable with Node and Deno. But I agree we have to make sure which one we want to recommend and take into account the security issues. 

> Perhaps, if you decide to use Node.js you could maintain the original Clarinet/Deno syntax Clarinet.test(... to keep all existing test-suites running...

Yes exactly, this is what I call the `compatibility layer` and it'll be crucial for the release of clarinet 2

**Reply by moodmosaic** [reply]
Sounds good to me, then.

**Comment by moodmosaic** [comment]
>I don't understand what you mean by that, could you briefly elaborate?

IMHO, existing tests written in Deno and TypeScript must run without modifications, even if you decide to switch to Node.js, because:
* trust in a test is generally established when the test is created. As time goes by, if tests are not modified but only appended, the test suite remains trustworthy. This concept is similar to the Open/Closed Principle, suggesting that a test suite should be open for extension, [but closed for modification​](https://blog.ploeh.dk/2013/04/02/why-trust-tests/).
* tests monitor the correctness of the smart contract, but there's no infinite regress of [tests testing tests​](https://blog.ploeh.dk/2013/04/02/why-trust-tests/)​...

**Comment by lgalabru** [comment]
The idea of writing clarity for testing clarity is not new, believe me [I thought about it and gave it a try](https://github.com/stacks-network/stacks-blockchain/issues/1167).
My conclusion was that clarity is factually, deliberately, a constrained language. It is designed to write smart contracts, not unit tests. 
Clarinet is certainly opinionated, and our line of thinking is the following: for testing smart contracts, we want an expressive, Turing complete language where we can reuse helpers or existing libraries and that let us set-up the chain in certain / extreme conditions, inter-placing certain transactions, etc. 
When writing unit tests, the goal is to test side effects of contract execution: we want to test for the impact of blocks on the chainstate and check events being emitted. You can't get this level of introspection in clarity.
We have a few examples of existing products / launched protocols, that were able to gracefully scale their clarity codebase + unit tests with this approach. Are these protocol broken, or struggling scaling because of this testing architecture? I try to stay in touch with developers, and I don't believe this is the case.

With this workstream, we want to start from a blank slate, and based on what we learned, bring smart contract testing to the next level both in terms of methodology and tooling - and I'm certain that fuzz testing is a piece of the puzzle. 
We will do our best to minimize the amount of work required for the transition from Clarinet 1 to Clarinet 2 for our developers, we could also probably open some pull requests in some selected repos to reduce the friction on the transition (which is something I did in the past), but constraining our design space with backward compatibility is probably undesirable. Last resort, Clarinet 1 will stay available, if developers want to stick with this version.

Foundry is successful today because before Foundry, there was Truffle and Hardhat. This ecosystem matured and competition pushed boundaries, with new approaches being started from scratch. 
Writing unit tests in Solidity (a Turing complete, expressive language) was possible in Truffle in 2017, if not earlier - this approach was not invented by Foundry.

**Reply by moodmosaic** [reply]
>writing clarity for testing clarity is not new [I thought about it and gave it a try](https://github.com/stacks-network/stacks-blockchain/issues/1167)

That's awesome 🚀 Can you share some details? I looked into that issue but there's nothing written inside wrt.

>for testing smart contracts, we want an expressive, Turing complete language where we can reuse helpers or existing libraries and that let us set-up the chain in certain / extreme conditions, inter-placing certain transactions, etc.

I could be wrong, it looks like you're mixing test *discovery*, *infrastructure*, *execution*, and the *assertion* mechanism(?) The way the sBTC mini protocol is being tested is to
* use Clarinet for test *discovery*, *infrastructure* and *execution*
* use Clarity language for the *assertion* mechanism
  * by that I mean that the [four phases](http://xunitpatterns.com/Four%20Phase%20Test.html) of a test are written in Clarity itself

I'm not part of the sBTC team so that's all I can observe from the outside. I apologize if I am wrong. However, the way sBTC does testing looks very efficient because they can easily move on to testing the *hard stuff* and staying sane:
* they can write tests that exercise the state of the contract at any given time (Invariant Tests) which can be done also in TypeScript but it's very, very, painful. I have some examples to show, in TypeScript, if anyone's interested.
* they can support tests with parameters (to effectively turn those into property and fuzz tests) since the PRNG can happen inside Clarinet/Deno. That's the same approach that `dapp` and `foundry` use; `dapp` uses QuickCheck, `foundry` uses proptest, Clarinet uses fast-check (not really, yet, but that's the best library I'm aware of that you can use and it's [the one I used](https://github.com/dubzzz/fast-check/discussions/2978)).

>When writing unit tests, the goal is to test side effects of contract execution: we want to test for the impact of blocks on the chainstate and check events being emitted. You can't get this level of introspection in clarity.

See my answer above.

>Foundry is successful today because before Foundry, there was Truffle and Hardhat.

Indeed, the core ideas behind foundry can be found in its predecessor [dapp.tools](https://github.com/dapphub/dapptools), though foundry comes with exceptional engineering, community and marketing on top of it (which is all great).

This can sound 'strict' but if you want even more robust and secure contracts, those contracts devs have to stay inside the contract language as much as possible (to master it as much as possible) -- I can't see how context switching between languages can help.

>We want to start from a blank slate, and based on what we learned, bring smart contract testing to the next level both in terms of methodology and tooling

Deno has no support for using transformer plugins (or at least there wasn't in 2022) and so runtime reflection was a real tricky one (I had to work with @Hookyns and create a custom, Deno friendlier, version of his `tst-reflect-transformer` plugins and use Node.js).

The end result was pretty impressive though: https://github.com/PromptECO/clarity-sequence/issues/4#issuecomment-1329303698

Perhaps if you work on Clarinet 2, you'd have to consider switching to Node.js, or somehow support transformer plugins in Deno...

**Reply by LNow** [reply]
With language versioning clearly defined on the code level (https://github.com/stacksgov/sips/pull/95#discussion_r1008914359) you would be able to define your own language version ie. `clarinet-v2` with your own set of functions, limitations and functionalities.
By doing so you would be able to write everything in Clarity and Clarity-ish language.

**Reply by hugoclrd** [reply]
Hey @LNow,
Although I think it could make some things easier to have the clarity version within contracts, I don't see how it relates to the tests. Could elaborate / give an example ?

**Reply by LNow** [reply]
If you have clarity version defined within the contract, you can define your own syntax for testing purpose. You can do it for simple unit-tests, fuzz-test, prop-based-test. 

Contract named `counter-contract`:
```clarity
(clarity-version "2.1")

(define-data-var counter uint u0)
(define-constant DEPLOYER tx-sender)

(define-public (increment (by uint))
  (begin
    (var-set counter (+ (var-get counter) by))
    (stx-transfer? u200 tx-sender DEPLOYER)
  )
)
```
Tests:
```clarity
(clarity-version "clarinet-2.1")

;; `test` - custom function that describes single test
;; (test (test-name) test-body)
(test (check-increment)
  (let (
    (sender1 'ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5)
    (sender2 'ST2JHG361ZXG51QTKY2NQCVBPPRRE2KZB1HR05NNC)
    (by u20)
    ;; `call-public` - another custom function, this time to define call of a public function
    (tx1 (call-public sender1 counter-contract increment by))
    (tx2 (call-public sender2 counter-contract increment by))
    ;; `mine-block` - another custom function; returns a tuple: { height: uint, receipts: (list receipt)}
    ;; where `receipts` is another tuple: { result: (response X Y), events: (list event)} 
    (block (mine-block (list tx1 tx2)))
    )
  
    (asserts! (is-eq ((get height block)) u17) (err "Invalid block height"))
    (asserts! (is-eq (len (get receipts block)) u2) (err "invalid number of receipts"))
    
    ;; `[x]` syntax could be defined as shortcut for `(unwrap-panic (element-at list X))`
    (asserts! (is-ok (get result (get receipts block)[0])) (err "Should succeed"))
    (asserts! (is-err (get result (get receipts block)[1])) (err "Should fail"))

    (asserts! (is-eq 
      (get events (get receipts block)[0])[0] 
      {type: "stx-transfer", from: sender1, to: 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM, amount: u200})
      (err "invalid event"))
    (asserts! (is-eq (len (get events (get receipts block)[1])) u0) (err "Failed transaction can't have events"))

  
  )
)
```

Above is just an example. You could define functions which allows you to access maps and variables without exposing them with read-only or public functions.
You could define set of helper functions that would make testing much easier. You could define constant values which are relevant only withing testing context.
You could do A LOT

**Reply by moodmosaic** [reply]
This could be an interesting addition to the sBTC Mini protocol [tests](https://github.com/Trust-Machines/stacks-sbtc/tree/main/sbtc-mini#unit-testing) which are written in Clarity (and discovered and executed by Clarinet).

**Reply by hugoclrd** [reply]
In the clarinet.toml you can already specify which clarity version to use for each contract, so from the clarinet pov it wouldn't change much to define at the contract level.

And clarinet relies directly on the clarity-vm, so _right now_ there is no mechanism to add a custom version of clarity

**Comment by igorsyl** [comment]
Any way we could remove any non-standard LISP syntax, for example:
```
{type: "stx-transfer"
```

@MarvinJanssen and I were discussing the idea to write LISP code to generate Clarity code as a DSL and having LISP strict syntax would be helpful. One application would be to have full LISP turing-complete environment for test cases. Other cases would be to leverage LISP macros and other matured tooling.

**Comment by hugoclrd** [comment]
The testing framework with node has been released 
https://www.npmjs.com/package/@hirosystems/clarinet-sdk

Deno testing has been removed in clarinet 2.0
